doc/compiler/optimizer/Inliner.md

```diff
@@ -0,0 +1,301 @@
+ 
+* **tryToGenerateILForMethod**
+* **inlineCallTargets**
+* **TR_TransformInlinedFunction::transform**
+* Stitching callee's IL into caller
+* Merging CFGs
+
+First, `inlineCallTarget2` calls `tryToInlineTrivialMethod`. `tryToInlineTrivialMethod` is a part of Inliner imbued with *some* knowledge of a target language. `tryToInlineTrivialMethod` is able to recognize certain *standard-library* methods of the target language and generate the IL for those avoiding the full inlining process of `inlineCallTarget2`. If `tryToInlineTrivialMethod` is successful, `inlineCallTarget2` returns immediately.
+
+Next, `inlineCallTarget2` makes a call to `tryToGenerateILForMethod` to generate callee's IL. Typically, the call is successful and inlining proceeds. 	
+
+Now that `inlineCallTarget2` got callee's IL it can *recursively* call `inlineCallTargets` to scout and inline possible calls in the callee. 
+
+Then, `inlineCallTarget2` initializes `TR_ParameterToArgumentMapper`. `TR_ParameterToArgumentMapper` as its name suggests needs to map *actual* arguments to formal *parameters*. `TR_ParameterToArgumentMapper` *can* directly reuse some of caller's arguments. Currently, the logic in `TR_ParameterToArgumentMapper` will reuse an incoming argument if 
+* it is a const, 
+* is is *not* modified by callee,
+* the argument is *not* used anywhere else, 
+* the argument is a *this* object and it is used *only* twice:
+	* to fetch a vft out of *this* object
+	* to be passed as *this* argument to the call. 
+	
+Otherwise, `TR_ParameterToArgumentMapper` either creates a fresh temporary to hold the argument value or reuses one of the available temps (see [Temp Sharing](#temp-sharing)) 
+
+
+`TR_TransformInlinedFunction` object is created and callee's IL transformed in the following ways: 
+ * replaces the uses of callee's parameters with the temps or args assigned by `TR_ParameterToArgumentMapper` via `transformNode`
+ * transforms callee's return node by 
+  * creating a new temporary and storing the return value into the temp
+  * replacing the result of the call (the use of the call node) in caller with the return value 
+  * ignoring the return value if the result of the call is not used. 
+
+
+After that, `inlineCallTarget2` embeds callee's IL into caller's. The IL includes all the stores required to move arguments into temporaries. Also, Inliner adds CFG callee's CFG (control flow graph) nodes to the caller's CFG.
+
+Next, if `TR_VirtualGuardSelection` of the calltarget dictates that an inlining guard is required for this call, Inliner invokes `addGuardForVirtual`. `addGuardForVirtual` adds a virtual snippet block and generates a configuration described in [Inline Guards](#inlining-guards). `addGuardForVirtual` uses `createVirtualGuard` to generate IL for the guard test (e.g. `ifacmpne`, `ificmpne`, etc). OMR Inliner currently supports single-condition guards only. @TODO Do we need more details here? 
```

Comment:
You mean the stuff `TR_VirtualGuardHeadMerger` does (merging two guards to go to the same cold block)? This is a little bit beyond what Inliner typically does with the guards. I did say 
> OMR Inliner currently supports single-condition guards only. 

since I remembered that Kai needed two tests to implement one guard. We ended up implementing it as a helper call since we didn't have infrastructure to do multiple tests.